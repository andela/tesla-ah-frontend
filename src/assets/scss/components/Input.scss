@import '../abstracts/variables';

.input {
  border: 1px solid $blue;
}

.input.error {
  border: 2px solid rgba(255, 0, 0, 0.733);
}

.input.pass {
  border: 1px solid rgba(0, 128, 0, 0.87);
}

small.error {
  color: rgba(255, 0, 0, 0.733) !important;
  font-weight: bold;
}

small.pass {
  color: rgba(0, 128, 0, 0.87) !important;
}
