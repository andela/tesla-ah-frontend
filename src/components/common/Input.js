/* eslint-disable jsx-a11y/label-has-associated-control */
/* eslint-disable jsx-a11y/label-has-for */

import React from 'react';
import { capitalize } from 'lodash';

import '../../assets/scss/components/Input.scss';

const Input = (props) => {
  const {
    invalid,
    shouldValidate,
    touched,
    inputtype,
    changed,
    elementConfig,
    value,
    formLabel,
    errorMessage,
  } = props;
  let inputElement = null;
  let validationHelper = null;
  const inputClasses = ['form-control input'];

  if (invalid && shouldValidate && touched) {
    inputClasses.push('error');
    validationHelper = (
      <small id="emailHelp" className="form-text text-muted error">
        {`${capitalize(formLabel)} ${errorMessage}`}
      </small>
    );
  }

  if (!invalid && shouldValidate && touched) {
    inputClasses.push('pass');
    validationHelper = (
      <small
        id="emailHelp"
        className="form-text text-muted pass"
        style={{ marginLeft: '5px' }}
      >
        <i className="fas fa-check" />
      </small>
    );
  }

  switch (inputtype) {
    case 'input':
      inputElement = (
        <input
          className={inputClasses.join(' ')}
          onChange={changed}
          {...elementConfig}
          value={value}
        />
      );
      break;
    case 'textarea':
      inputElement = (
        <textarea
          className={inputClasses.join(' ')}
          onChange={changed}
          {...elementConfig}
          value={value}
        />
      );
      break;
    case 'select':
      inputElement = (
        <select
          className={inputClasses.join(' ')}
          onChange={changed}
          value={value}
        >
          {elementConfig.options.map(option => (
            <option key={option.value} value={option.value}>
              {option.displayValue}
            </option>
          ))}
        </select>
      );
      break;
    default:
      inputElement = (
        <input
          className={inputClasses.join(' ')}
          onChange={changed}
          {...elementConfig}
          value={value}
        />
      );
  }

  return (
    <div className="form-group d-flex flex-column align-items-start">
      <label htmlFor="exampleInputPassword1">{capitalize(formLabel)}</label>
      {inputElement}
      {validationHelper}
    </div>
  );
};

export default Input;
