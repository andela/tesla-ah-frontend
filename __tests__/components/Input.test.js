/* eslint-disable no-undef */
import React from 'react';
import { shallow } from 'enzyme';
import Input from '../../src/components/common/Input';

const renderInput = (args) => {
  const defaultProps = {
    invalid: '',
    shouldValidate: '',
    inputtype: '',
    changed: jest.fn(),
    elementConfig: {},
    value: '',
    formLabel: '',
    errorMessage: '',
  };
  const props = { ...defaultProps, ...args };

  return shallow(<Input {...props} />);
};

it('renders Input component', () => {
  const wrapper = renderInput();
  expect(wrapper).toMatchSnapshot();
});

describe('Test supported input field types', () => {
  it('renders the input field by default when no inputtype is specified', () => {
    const wrapper = renderInput({ inputtype: '' });
    expect(wrapper.find('input').length).toBe(1);
  });

  it('renders the input field', () => {
    const wrapper = renderInput({ inputtype: 'input' });
    expect(wrapper.find('input').length).toBe(1);
  });

  it('renders the textarea field', () => {
    const wrapper = renderInput({ inputtype: 'textarea' });
    expect(wrapper.find('textarea').length).toBe(1);
  });

  it('renders the select field', () => {
    const options = [
      { value: '', displayValue: 'Pick a gender' },
      { value: 'M', displayValue: 'Male' },
      { value: 'F', displayValue: 'Female' },
    ];
    const wrapper = renderInput({
      inputtype: 'select',
      elementConfig: { options },
    });
    expect(wrapper.find('select').length).toBe(1);
    expect(wrapper.find('option').length).toBe(3);
  });
});

it('should have pass outline if valid', () => {
  const wrapper = renderInput({
    inputtype: 'input',
    invalid: false,
    shouldValidate: true,
    touched: true,
  });
  expect(wrapper.find('small.pass').length).toBe(1);
});

it('should have error outline if invalid', () => {
  const wrapper = renderInput({
    inputtype: 'input',
    invalid: true,
    shouldValidate: true,
    touched: true,
  });
  expect(wrapper.find('small.error').length).toBe(1);
});
