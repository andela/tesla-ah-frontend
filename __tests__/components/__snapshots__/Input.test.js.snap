// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Input component 1`] = `ShallowWrapper {}`;
